### DISCUSSION TITLE
About ResponseBuilder

### DISCUSSION THREAD

**Discussion by SuenagaRyuya** [aws-powertools/powertools-lambda-python#2994] [General]
Hi!

I have a question about ResponseBuilder class in `aws_lambda_powertools.event_handler.api_gateway`.

I just want to use `build` method in this class to return response from api gateway.
But from version 2, we can't use the method because `aws_lambda_powertools.shared.header_serializer` raise `NotImplementedError`.

So if I use this function, should I use powertools version 1?

Thank you!

**Comment by leandrodamascena** [comment]
Hello @SuenagaRyuya!

Can you please give me an example of what you're achieving when using the `build` method? Can't you use the `Response` class to build the response to APIGateway?

I appreciate if you can provide a code example.

Thank you.

**Reply by SuenagaRyuya** [reply]
Hello @leandrodamascena !
Thank you for replying!

I try to do like this!
```python
from aws_lambda_powertools.event_handler.api_gateway import Response, ResponseBuilder

def lambda_handler(event, context):
    # do something...

    response = Response(status_code=200, ....)

    response_builder = ResponseBuilder(response)
    
    return response_builder.build()
```

Is it correct?

I want to implement microservice api with api gateway and aws lambda.
I think `APIGatewayRestResolver` is for monolithic architecture.

If there are other way to implement, could you let me know?

Thank you!

**Reply by leandrodamascena** [reply]
Hello @SuenagaRyuya! I see a few points here and let me try to address them.

1 - The short answer is: `Response` is an object and in Python you can access the `__dict__` property, which is a json representation of this object. You can use this code to get what you want.

```python
from aws_lambda_powertools.event_handler import Response

def lambda_handler(event, context):
    
    response = Response(status_code=200, body="Powertools", ...).__dict__

    return response
```

2 - You mention that you are using microservice architecture and `APIGatewayRestResolver` is just for the monolithic architecture. I think this is not like this, you still can use `APIGatewayRestResolver` in your microservices architecture. If you use the `APIGatewayRestResolver` you can take advantage of many things like enabling debug mode, easily creating binary responses, configuring CORS, accessing request details, and others. Also, support for Middleware is coming soon (https://github.com/aws-powertools/powertools-lambda-python/pull/2917) which is an important feature.
I see many reasons to use the `APIGatewayRestResolver` in a microservice/micro-function architecture. 

Let me know if you want to share some ideas and we can try to create something together to show the benefits of using this resolver.

Thank you.

**Reply by SuenagaRyuya** [reply]
Hello @leandrodamascena!
Thank you for giving me a example solving my problem!
I think I can implement this function because of the example!

And I don't have enough knowledge about architecture.
So I try to learn this subject more and think about architecture.

Thank you for teaching me the way from the bottom of my heart!
Let you know if there are some ideas!

Thank you!

**Reply by leandrodamascena** [reply]
Hello @SuenagaRyuya! That's really amazing to hear! I'm glad I could help you solve this problem.

We have a Discord server where we talk more about Powertools and how it can help you. If you're not part of it, feel free to join - https://discord.gg/B8zZKbbyET.

**Comment by leandrodamascena** [comment]
Closing this discussion as completed!

If needed, please reopen.
